### ISSUE TITLE
Time embedding scale + 1 / classifier-free guidance

### ISSUE THREAD

**Issue by ajrheng** [lucidrains/denoising-diffusion-pytorch#120]
According to #77, FiLM conditioning is used for the time embeddings instead of just simple shifting. May I know what is the reason for the +1 added to the scaling?
`x = x * (scale + 1) + shift
`

On a related note, if I wanted to implement [classifier-free guidance](https://arxiv.org/abs/2207.12598) to do conditional generation, would you recommend implementing the class-conditioning embedding with FiLM in the same way as time? Would appreciate if anyone could point me to a github implemention of classifier-free guidance for DDPM--looks like the original paper does not have open-source code. 

Thank you!

**Comment by lucidrains** [comment]
@ajrheng Hi Alvin, so that type of film conditioning was actually how FiLM conditioning was implemented in [GLIDE](https://github.com/openai/glide-text2im). for classifier free guidance, you can look at a number of repositories, from GLIDE to [dalle2](https://github.com/lucidrains/DALLE2-pytorch) to [imagen](https://github.com/lucidrains/imagen-pytorch)

the rationale explained to me was that the unet is operating in a range of -1 to 1, and doing a +1 shifted the scale to be positive. but i think having some ablation in the paper would be more convincing argument for its use

**Comment by lucidrains** [comment]
@ajrheng Hi Alvin, so the rationale for the +1 is to basically make sure the FiLM initial output is ~ identity function to start off with, assuming that the scale and shift will be centered around 0

for examples of classifier free guidance, just study [this function](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L1467)

**Comment by ajrheng** [comment]
@lucidrains Thanks Phil for the quick response.

1) Is there a reason why we want FiLM to be identity at the start? Was it found to work better empirically perhaps? 
2) Thanks for the link. That's equation 6 of the paper I presume, where you combine conditional and unconditional outputs. My question was exactly how should we implement the class-conditioning signal. It seems for that repo the signal is text tokens and it looks rather involved. For a simple class signal (like CIFAR10 classes), would it be appropriate to process it sinusoidally (like time) and just append it with time signal before passing into FiLM?

**Comment by lucidrains** [comment]
1. yea, it is just a better initialization
2. i would just give each class its own embedding, and do the conditional dropout by swapping it out for a null embedding

maybe if i find some time this week, i'll build it out in a separate file

**Comment by ajrheng** [comment]
Are you suggesting to fix the class embeddings and not train them? Presumably these embeddings are still inserted into the UNet layers in a FiLM manner?

I would greatly appreciate if you have the time to build out a demo. Thank you very much, Phil!

**Comment by lucidrains** [comment]
@ajrheng ok i'll build it today, i think it'll help a lot of students

**Comment by ajrheng** [comment]
@lucidrains Thank you! That will indeed be extremely helpful. Please let me know where I can find it when you are finished!

**Comment by lucidrains** [comment]
@ajrheng oh sorry, didn't get to it today 😢 and i'm going to go see Wakanda Forever

tomorrow morning!

**Comment by ajrheng** [comment]
No problem. Enjoy the movie! :)

**Comment by lucidrains** [comment]
@ajrheng hey Alvin, let me know if https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/classifier_free_guidance.py is helpful (example script at the bottom)

**Comment by ajrheng** [comment]
Thank you so much Phil! This will help with my project so much. I will get about testing it immediately. Will close this issue now!

**Comment by Mat-Po** [comment]
@ajrheng Hi Alvin! I am trying to follow your path toward class conditioning diffusion models here. I saw the [file](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/classifier_free_guidance.py) in the previous comment by Phil which was super useful. I was trying to adapt the Trainer class here to speed up the training but i am having some difficulties. did you menage to speed up your training in this or in any other ways?

I also saw [this](https://github.com/lucidrains/classifier-free-guidance-pytorch) other repo from Phil but i felt that in my case, (conditioning on 6 classes) is much better the solution suggested here by Phil. Have you tried it or have any comment on this?

**Comment by ajrheng** [comment]
@Mat-Po Hi Matteo, I don't have too many insights on speeding up the code. I personally rewrote some of my code so I could use `DataParallel` over the `Accelerate` library but that is just because `DataParallel` works better for my debugging workflow. Don't think I saw any speed increase. 

One area I did notice is that if you're working with large image datasets (high resolutions and lots of files), the loading of data into the dataloader can be slow. A lot of the time I observed my GPU was idle because it was waiting for the next batch to be loaded. I'm not too sure the best way to resolve this though.

I wasn't aware Phil has a new repo on classifier-free guidance but personally testing the one implemented in this repo works perfectly well for me for CIFAR10 (10 classes), so I don't think you would have any issues.

**Comment by Mat-Po** [comment]
@ajrheng Hi Alvin,
I usually did not notice any great advantage in using  pytorch's `DataParallel` , but that's also because is not the best option. I would suggest you the `DistributedDataParallel` as training strategy (which is what accelerate and pytorch lightning (pl) use, as base mode). With that, or with the direct help of Acelerate/pl, i usually notice a big speed up. However adjusting the code is not always trivial.
